src/uri.rs

```diff
@@ -207,13 +250,19 @@ impl FromStr for Uri {
     type Err = Error;
 
     fn from_str(s: &str) -> Result<Uri, Error> {
-        Uri::new(s)
+        Uri::new(InternalUri::from(s))
     }
 }
 
 impl From<Url> for Uri {
     fn from(url: Url) -> Uri {
-        Uri::new(url.as_str()).expect("Uri::From<Url> failed")
+        Uri::new(InternalUri::from(url.as_str())).expect("Uri::From<Url> failed")
+    }
+}
+
+impl From<MemSlice> for Uri {
+    fn from(m: MemSlice) -> Uri {
+        Uri::new(InternalUri::from(m)).expect("Uri::From<MemSlice> failed")
```

Comment:
A `MemSlice` comes from receiving bytes from the socket. It has gone through `httparse`, which made sure there weren't radically invalid bytes here (just that it is a `str` without spaces), but we don't know if it is actually a well-formed `Uri`. So, we can't really provide the `From<MemSlice>` implementation, here.